crates/matrix-sdk/src/oidc/impl/mod.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for that specific language governing permissions and
+// limitations under the License.
+
+//! Trait for defining an implementation for an OIDC backend.
+//!
+//! Used mostly for testing purposes.
+
+use mas_oidc_client::{
+    requests::authorization_code::{AuthorizationRequestData, AuthorizationValidationData},
+    types::{
+        client_credentials::ClientCredentials,
+        iana::oauth::OAuthTokenTypeHint,
+        oidc::VerifiedProviderMetadata,
+        registration::{ClientRegistrationResponse, VerifiedClientMetadata},
+        IdToken,
+    },
+};
+use url::Url;
+
+use super::{AuthorizationCode, OidcError, SessionTokens};
+
+mod server;
+pub(crate) use server::*;
+
+#[cfg(test)]
+mod test;
+#[cfg(test)]
+pub(crate) use test::*;
+
+pub(super) struct RefreshedSessionTokens {
+    pub access_token: String,
+    pub refresh_token: Option<String>,
+}
+
+#[async_trait::async_trait]
+pub(super) trait OidcImpl: std::fmt::Debug + Send + Sync {
```

Comment:
Shouldn't it be named `OidcBackend` as that's what corresponding bindings are called?

Reply:
Sure